docs/proposal/20220929-ipam-support.md

```diff
@@ -0,0 +1,422 @@
+> ip-address-manager with external systems.
+>
+> All on-premise providers have a need for IP Address Management, since they
+> can't leverage SDN features that are available in clouds such as AWS, Azure or
+> GCP. We therefore propose to add an API contract to CAPI, which allows
+> infrastructure providers to integrate with different IPAM systems. A similar
+> approach to Kubernetes' PersistentVolumes should be used where infrastructure
+> providers create Claims that reference a specific IP Pool, and IPAM providers
+> fulfill claims to Pools managed by them.
+
+### Goals
+
+- Implement support for the IPAM API contract in CAPV
+- Support both IPV4 and IPv6
+- Support dual-stack
+- Allow dynamically allocating and releasing addresses
+- Allow running multiple IPAM providers in parallel
+- Support running any IPAM provider
+
+### Non-Goals/Future Work
+
+- Configuring nameservers from Pool
+- Handling updates to the devices on VSphereMachine objects
+- Handling updates to the fromPools list on individual devices
+- User direct manipulation of IPAddressClaims
+- IP Address Management – this is the job of the IPAM provider
+- Support for specific IPs for specific machines - a machine will receive a
+  random IP in the pool.
+
+## Proposal
+
+Following the contract specified by the IPAM integration proposal we plan to:
+
+- Add the fromPools field to the spec.network.devices section of the
+  VSphereMachine and VSphereMachineTemplate. The fromPools field will allow a
+  user to specify that the network device IPAddress should be claimed from an
+  IPPool.
+- If a VSphereMachine has been specified to claim an IP Address from an IPPool,
+  then CAPV will create an IPAddressClaim, which the IPAM provider will use to
+  allocate an IP Address. This IPAddress can then be used by the
+  VSphereMachine.
+
+### User Stories
+
+#### Story 1
+
+As a user, I want to be able to create a VSphereMachine where the IP Address
+for the machine comes from an IPAM Provider instead of DHCP.
+
+#### Story 2
+
+As a user I want to be able to allocate either IPv4 or IPv6 addresses for my
+VSphereMachine.
+
+#### Story 3
+
+As a user I want to be able to release my IP Address allocation when the
+VSphereMachine is deleted for future VSphereMachines to use.
+
+### Implementation Details/Notes/Constraints
+
+#### API Changes
+
+Adding `fromPools`(type []corev1.TypedLocalObjectReference) to the VSphereMachineTemplate
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachineTemplate
+metadata:
+  name: example
+  namespace: vsphere-site1
+spec:
+  template:
+    spec:
+      cloneMode: FullClone
+      numCPUs: 8
+      memoryMiB: 8192
+      diskGiB: 45
+      network:
+        devices:
+        - dhcp4: false
+          dhcp6: false
+          networkName: VM Network
+          fromPools:
+            # reference to the pool
+            - group: ipam.cluster.x-k8s.io/v1alpha1

```

Comment:
This should either be _only_ the group, aka. `ipam.cluster.x-k8s.io`, or be renamed to `apiVersion`.

Reply:
Good catch, this is a mistake in the proposal. The field is not called group, it's called `apiGroup`. Update coming.